### ISSUE TITLE
coqc -quick fails to report an error message when a definition has an error

### ISSUE THREAD

**Issue by tchajed** [coq/coq#6707]
#### Version

The Coq Proof Assistant, version 8.8+alpha (February 2018)
compiled on Feb 7 2018 11:21:28 with OCaml 4.06.0

#### Operating system

OS X 10.11

#### Description of the problem

I don't have a good summarization of what's wrong, but `coqc -quick` repeatedly fails to report a sane error message when I have an error in a definition and then have a theorem later on. Here are a bunch of examples. Every one of these files results in a type error in the definition of `foo` when compiled with `coqc`.

I did a bisect of the behavior in `Test1.v` and this is a regression since v8.5 due to 4abb41d008fc754f21916dcac9cce49f2d04dd6d. However, I noticed that in early versions of Coq I would sometimes get both errors (that `tt` is not a function and that `foo` is undefined), which would be the ideal behavior.

Note that `Test2.v` and `Test5.v` are particularly concerning because in the first case `foo` is silently dropped and in the second `foo` and the theorem `a_foo` are dropped; `coqc` exits with success for both files.

```coq
(* Test1.v *)
Definition foo := tt tt.

Theorem a_foo : foo.
Admitted.
````

```
$ coqc -quick Test1.v
File "./Test1.v", line 4, characters 16-19:
Error: The reference foo was not found in the current environment.

```
```coq
(* Test2.v *)
Definition bar := True.
Definition foo := tt tt.

Theorem a_foo : bar.
Admitted.
````

```
$ coqc -quick Test2.v
```
```coq
(* Test3.v *)
Definition bar := True.
Definition foo := tt tt.

Theorem a_foo : foo.
Admitted.
````

```
$ coqc -quick Test3.v
File "./Test3.v", line 5, characters 16-19:
Error: The reference foo was not found in the current environment.

```
```coq
(* Test4.v *)
Definition bar := True.
Definition foo := tt tt.

Theorem a_bar : bar.
Admitted.

Theorem a_foo : foo.
Admitted.
````

```
$ coqc -quick Test4.v
File "./Test4.v", line 8, characters 16-19:
Error: The reference foo was not found in the current environment.

```
```coq
(* Test5.v *)
Definition bar := True.
Definition foo := tt tt.

Theorem a_foo : foo.
Admitted.

Theorem a_bar : bar.
Admitted.
````

```
$ coqc -quick Test5.v
```
```coq
(* Test6.v *)
Definition foo := tt tt.

Theorem a_foo : foo.
Admitted.

Theorem a_bar : bar.
Admitted.
````

```
$ coqc -quick Test6.v
File "./Test6.v", line 7, characters 16-19:
Error: The reference bar was not found in the current environment.

```
```coq
(* Test7.v *)
Definition foo := tt tt.

Theorem a_bar : bar.
Admitted.
````

```
$ coqc -quick Test7.v
File "./Test7.v", line 4, characters 16-19:
Error: The reference bar was not found in the current environment.

```

**Comment by Zimmi48** [comment]
@tchajed Did you check whether #6262 fixes that?

**Comment by ejgallego** [comment]
#6262 won't fix that. This is a subtle issue, my first guess is that the resilient error code is maybe kicking-in so no fatal error is generated.

**Comment by tchajed** [comment]
Yeah, #6262 doesn't make a difference here.

**Comment by ejgallego** [comment]
So indeed for the two cases that exit without error, the error recovery mechanism introduced in 8.6 is making them non fatal. The right fix is to disable this feature in non-interactive mode.

**Comment by ejgallego** [comment]
Try this:
```
modified   toplevel/coqtop.ml
@@ -354,6 +354,8 @@ let compile ~time ~verbosely ~f_in ~f_out =
         | None -> long_f_dot_v ^ "io"
         | Some f -> ensure_vio long_f_dot_v f in
 
+      let stm_options = { stm_options with Stm.AsyncOpts.async_proofs_cmd_error_resilience = false } in
+
       let doc, sid = Stm.(new_doc
           { doc_type = VioDoc long_f_dot_vio;
             iload_path;
```

**Comment by ejgallego** [comment]
Keep in mind that if we wouldn't disable the error resiliency code, 

In 8.7 a quick try can be done passing the option `-async-proofs-command-error-resilience false`.

**Comment by ejgallego** [comment]
Well, even setting this flag won't make the STM generate an exception. it is a bit hard then as `coqtop` has no way of understanding errors reliably for now other than by getting an exception.

IMHO this is a bug in the error resilience code, @gares , why even after setting the flag to false no exception is generated?

**Comment by ejgallego** [comment]
There is another possibility and indeed it that we are facing #5479 again. @tchajed , what is the exit code for Test2.v in 8.5 / versions that print the message?

**Comment by ejgallego** [comment]
For people curious about this, I debug what's really happening under the hood using SerAPI:
```
(Add () "Definition foo := tt tt. Theorem a_foo : foo. Admitted.")
Finish
```
gives the full info about what Coq is thinking.

**Comment by tchajed** [comment]
Turning off command error resilience (with either `coqtop -async-proofs-command-error-resilience no -quick -compile Test2.v` or the diff you gave) fixes all the examples.

**Comment by tchajed** [comment]
Coq v8.5 exits with status 1 as expected.

**Comment by ejgallego** [comment]
Good to hear. I'll prepare a fix, which should be backported to 8.7/8.6 too IMHO.

> Coq v8.5 exits with status 1 as expected.

The right commit to test would be indeed after the merge of the command resilience patch, in that case, I guess that we would be in a situation where the error message is displayed, but `coqc` exists with 0.

The commit you have pointed out removed the message, but the real culprit should be around ff67a511a358ada3daefea0839e18d474531e13d
